File: code/jvm/service/src/main/java/pt/isel/services/messages/MessageService.java
package pt.isel.services.messages;

import jakarta.inject.Named;
import pt.isel.domain.channels.AccessType;
import pt.isel.domain.channels.Channel;
import pt.isel.domain.common.Either;
import pt.isel.domain.common.MessageError;
import pt.isel.domain.messages.Message;
import pt.isel.domain.messages.UpdatedMessage;
import pt.isel.domain.users.UserInfo;
import pt.isel.repositories.Transaction;
import pt.isel.repositories.TransactionManager;
import pt.isel.services.common.RateLimiter;

import java.time.Clock;
import java.time.Duration;
import java.time.LocalDateTime;
import java.util.List;

@Named
public class MessageService {
    private final TransactionManager trxManager;
    private final MessageEventService messageEventService;
    private final RateLimiter rateLimiter;
    private final Clock clock;

    public MessageService(TransactionManager trxManager, MessageEventService messageEventService, RateLimiter rateLimiter, Clock clock) {
        this.trxManager = trxManager;
        this.messageEventService = messageEventService;
        this.rateLimiter = rateLimiter;
        this.clock = clock;
    }

    public Either<MessageError, Message> createMessage(String content, Long userId, Long channelId) {
        if (rateLimiter.isRateLimited("message", userId.toString(), 30, Duration.ofMinutes(1))) {
            return Either.failure(new MessageError.RateLimitExceeded());
        }

        if (content == null || content.isBlank()) return Either.failure(new MessageError.EmptyMessage());
        if (content.length() > 1000) return Either.failure(new MessageError.InvalidMessageLength());

        var result = trxManager.run(trx -> checkUserCanPostMessage(trx, userId, channelId)
                .flatMap(pair -> {
                    var message = trx.repoMessages().create(content, pair.userInfo(), pair.channel(), LocalDateTime.now(clock));
                    return Either.success(message);
                }));

        if (result instanceof Either.Right<MessageError, Message>(var message)) {
            messageEventService.broadcastMessage(channelId, new UpdatedMessage.NewMessage(message));
        }

        return result;
    }

    public Either<MessageError, List<Message>> getMessagesInChannel(Long userId, Long channelId, int limit, int offset) {
        if (limit <= 0) return Either.failure(new MessageError.InvalidLimit());
        if (offset < 0) return Either.failure(new MessageError.InvalidOffset());

        return trxManager.run(trx -> {
            Channel channel = trx.repoChannels().findById(channelId);
            if (channel == null) return Either.failure(new MessageError.ChannelNotFound());

            if (trx.repoMemberships().findUserInChannel(userId, channelId) == null) {
                return Either.failure(new MessageError.UserNotInChannel());
            }

            List<Message> messages = trx.repoMessages().findAllInChannel(channel, limit, offset);
            return Either.success(messages);
        });
    }

    private Either<MessageError, UserChannelPair> checkUserCanPostMessage(Transaction trx, Long userId, Long channelId) {
        var user = trx.repoUsers().findById(userId);
        if (user == null) return Either.failure(new MessageError.UserNotFound());

        var channel = trx.repoChannels().findById(channelId);
        if (channel == null) return Either.failure(new MessageError.ChannelNotFound());

        var membership = trx.repoMemberships().findUserInChannel(user.id(), channel.id());
        if (membership == null) return Either.failure(new MessageError.UserNotInChannel());

        if (membership.accessType() != AccessType.READ_WRITE) {
            return Either.failure(new MessageError.UserNotAuthorized());
        }

        UserInfo userInfo = new UserInfo(user.id(), user.username());
        return Either.success(new UserChannelPair(userInfo, channel));
    }

    private record UserChannelPair(UserInfo userInfo, Channel channel) {
    }
}

File: code/jvm/service/src/main/kotlin/pt/isel/services/messages/MessageService.kt
package pt.isel.services.messages

import jakarta.inject.Named
import pt.isel.domain.channels.AccessType
import pt.isel.domain.channels.Channel
import pt.isel.domain.messages.Message
import pt.isel.domain.messages.NewMessage
import pt.isel.domain.users.UserInfo
import pt.isel.repositories.Transaction
import pt.isel.repositories.TransactionManager
import pt.isel.services.common.Either
import pt.isel.services.common.Failure
import pt.isel.services.common.MessageError
import pt.isel.services.common.Success
import pt.isel.services.common.failure
import pt.isel.services.common.success

@Named
class MessageService(
    private val trxManager: TransactionManager,
    private val messageEventService: MessageEventService,
) {
    fun createMessage(
        content: String,
        userId: Long,
        channelId: Long,
    ): Either<MessageError, Message> {
        if (content.isBlank()) return failure(MessageError.EmptyMessage)
        if (content.length !in 1..1000) return failure(MessageError.InvalidMessageLength)

        val result =
            trxManager.run {
                when (val checkResult = checkUserCanPostMessage(userId, channelId)) {
                    is Failure -> {
                        checkResult
                    }

                    is Success -> {
                        val (userInfo, channel) = checkResult.value
                        val message = repoMessages.create(content, userInfo, channel)
                        success(message)
                    }
                }
            }

        if (result is Success) {
            messageEventService.broadcastMessage(channelId, NewMessage(result.value))
        }

        return result
    }

    fun getMessagesInChannel(
        userId: Long,
        channelId: Long,
        limit: Int = 50,
        offset: Int = 0,
    ): Either<MessageError, List<Message>> {
        if (limit <= 0) return failure(MessageError.InvalidLimit)
        if (offset < 0) return failure(MessageError.InvalidOffset)

        return trxManager.run {
            val channel =
                repoChannels.findById(channelId) ?: return@run failure(MessageError.ChannelNotFound)

            repoMemberships.findUserInChannel(userId, channelId)
                ?: return@run failure(MessageError.UserNotInChannel)

            val messages = repoMessages.findAllInChannel(channel, limit, offset)

            success(messages)
        }
    }

    private fun Transaction.checkUserCanPostMessage(
        userId: Long,
        channelId: Long,
    ): Either<MessageError, Pair<UserInfo, Channel>> {
        val user = repoUsers.findById(userId) ?: return failure(MessageError.UserNotFound)
        val channel =
            repoChannels.findById(channelId) ?: return failure(MessageError.ChannelNotFound)
        val membership =
            repoMemberships.findUserInChannel(user.id, channel.id)
                ?: return failure(MessageError.UserNotInChannel)

        if (membership.accessType != AccessType.READ_WRITE) {
            return failure(MessageError.UserNotAuthorized)
        }
        val userInfo = UserInfo(user.id, user.username)
        return success(userInfo to channel)
    }
}
